src/net.c

```diff
@@ -1143,11 +1166,131 @@ coap_send(coap_session_t *session, coap_pdu_t *pdu) {
     memset(&block, 0, sizeof(block));
   }
 
+#if COAP_Q_BLOCK_SUPPORT
+  /* Indicate support for Q-Block if appropriate */
+  if (session->block_mode & COAP_BLOCK_TRY_Q_BLOCK &&
+      session->type == COAP_SESSION_TYPE_CLIENT &&
+      COAP_PDU_IS_REQUEST(pdu)) {
+    /*
+     * When the pass / fail response for Q-Block is received, this PDU will
+     * get transmitted.
+     */
+    return coap_block_test_q_block(session, pdu);
+  }
+#endif /* COAP_Q_BLOCK_SUPPORT */
+
+#if !defined(WITH_LWIP)
+  unsigned ref;
+
+  /* TODO for LwIP */
+  /*
+   * Need to wait for Q-Block test to get out and response back before
+   * continuing
+   */
+  ref = ++session->ref;
```

Comment:
This should be ref = coap_session_reference(session);

Reply:
Hmm, coap_session_reference() returns a pointer to the session, not the reference counter.  so, the code would have to be
````
coap_session_reference(session);
ref = session->ref;
````
which does open up a timing window when multi-threading support is done.